GVFS/GVFS.Common/GVFSEnlistmentStatistics.cs

```diff
@@ -0,0 +1,189 @@
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a '/', then that means it falls within the root
+        /// However if a directory's path doesn't contain a '/', it counts towards itself for hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        private void CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, ref int count, string parentDirectory, bool isFile)
+        {
+            foreach (string path in paths)
+            {
+                // Ensure all paths have the same slashes
+                string formattedPath = path.Replace('\\', '/');
+                if (formattedPath.StartsWith("/"))
+                {
+                    formattedPath = formattedPath.Substring(1);
+                }
+
+                if (formattedPath.StartsWith(parentDirectory))
+                {
+                    count++;
+
+                    if (isFile)
+                    {
+                        string topDir = this.ParseTopDirectory(this.TrimDirectoryTarget(formattedPath, parentDirectory));
+                        directoryTracking[topDir] = directoryTracking.ContainsKey(topDir) ? directoryTracking[topDir] + 1 : 1;
+                    }
+                    else
+                    {
+                        if (!parentDirectory.Equals(formattedPath))
+                        {
+                            string topDir = this.TrimDirectoryTarget(formattedPath, parentDirectory);
+                            if (topDir.IndexOf('/') != -1)
+                            {
+                                topDir = this.ParseTopDirectory(topDir);
+                            }
+
+                            directoryTracking[topDir] = directoryTracking.ContainsKey(topDir) ? directoryTracking[topDir] + 1 : 1;
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Format a given path by trimming off the leading part containing the target directory
+        /// </summary>
+        /// <param name="toTrim">The path to trim the original directory off of</param>
+        /// <returns>The newly formatted path without that initial directory</returns>
+        private string TrimDirectoryTarget(string path, string directoryTarget)
+        {
+            if (directoryTarget.Length == 0)
+            {
+                return path;
+            }
+            else
+            {
+                return path.Substring(directoryTarget.Length + 1);

```

Comment:
Is it possible to have directoryTarget longer than path?

Reply:
In theory it would be possible yes. I'll add a check and write a unit test for good measure!

Reply:
Upon further investigation, this is not possible. The only way you're even calling this function is when `formattedPath.StartsWith(parentDirectory)` is true, and therefor the length of path is >= the length of directoryTarget